File: python/features/steps/commit_steps.py
"""Behave steps for kbs commit."""

from __future__ import annotations

import subprocess

from pathlib import Path

from behave import then

from features.steps.shared import load_project_directory


@then("project/issues should be committed to git")
def then_project_issues_committed_to_git(context: object) -> None:
    working_directory = Path(context.working_directory).resolve()
    project_dir = load_project_directory(context)
    issues_path = (project_dir / "issues").resolve()
    relative_issues_path = issues_path.relative_to(working_directory).as_posix()
    result = subprocess.run(
        ["git", "status", "--porcelain", "--", relative_issues_path],
        cwd=working_directory,
        capture_output=True,
        text=True,
        check=False,
    )
    assert result.returncode == 0, result.stderr or result.stdout
    assert not result.stdout.strip(), result.stdout

File: rust/features/steps/commit_steps.rs
use std::path::PathBuf;
use std::process::Command;

use cucumber::then;

use kanbus::file_io::load_project_directory;

use crate::step_definitions::initialization_steps::KanbusWorld;

fn load_project_dir(world: &KanbusWorld) -> PathBuf {
    let cwd = world.working_directory.as_ref().expect("cwd");
    load_project_directory(cwd).expect("project dir")
}

#[then("project/issues should be committed to git")]
fn then_project_issues_committed_to_git(world: &mut KanbusWorld) {
    let cwd = world.working_directory.as_ref().expect("cwd");
    let cwd = cwd.canonicalize().unwrap_or_else(|_| cwd.clone());
    let project_dir = load_project_dir(world);
    let issues_path = project_dir.join("issues");
    let issues_path = issues_path.canonicalize().unwrap_or(issues_path);
    let relative_issues_path = issues_path
        .strip_prefix(&cwd)
        .expect("issues path under repo root")
        .to_string_lossy()
        .replace('\\', "/");

    let output = Command::new("git")
        .args(["status", "--porcelain", "--", &relative_issues_path])
        .current_dir(&cwd)
        .output()
        .expect("git status");

    assert!(
        output.status.success(),
        "{}",
        String::from_utf8_lossy(&output.stderr)
    );
    let stdout = String::from_utf8_lossy(&output.stdout);
    assert!(stdout.trim().is_empty(), "{stdout}");
}
